altair/view.py

```diff
@@ -1,116 +1,239 @@
-#         with output:
-#             v.render()
-#     render_button.on_click(on_render)
-#     return hbox
+class AltairWidget(object):
+    _shelves = ('x', 'y', 'row', 'col', 'shape', 'color', 'size', 'detail', 'text')
 
+    _shelf_types = {
+        'x': X,
+        'y': Y,
+        'row': Row,
+        'col': Col,
+        'shape': Shape,
+        'color': Color,
+        'detail': Detail,
+        'text': Text,
+        'size': Size
+    }
+
+    def __init__(self, viz, shelves=_shelves, renderer='lightning'):
+        use_renderer(renderer)
+
+        self.viz = viz
+        self.shelves = shelves
+
+        self._shelf_view_factories = {
+            'x': self._create_shelf_view,
+            'y': self._create_shelf_view,
+            'row': self._create_shelf_view,
+            'col': self._create_shelf_view,
+            'shape': self._create_shelf_view,
+            'color': self._create_shelf_view,
+            'detail': self._create_shelf_view,
+            'text': self._create_shelf_view,
+            'size': self._create_shelf_view
+        }
+
+        self.widget = w.HBox()
+
+        # Create output area
+        self.output = w.HBox()
+        self.output.children = []
+
+        # Create widget
+        self.controls = self._create_left_pane()
+
+        self.widget.children = [self.controls, self.output]
+
+    def render(self):
+        return self.widget
+
+    def on_render_viz(self, *args):
+        self._clear_output()
+
+        to_display = w.Textarea(value=str(self.viz))
+
+        # TODO: how to display properly?
+        # to_display = w.Output()
+        # with to_display:
+        #     self.viz.render()
+
+        self.output.children = [to_display]
+
+        return self.render()
+
+    def _clear_output(self):
+        for c in self.output.children:
+            c.visible = False
+            method = getattr(c, 'clear_output', None)
+            if method is not None:
+                method()
+
+        self.output.children = []
+
+    def _create_left_pane(self):
+        # Create types of viz hbox
+        viz_types_hbox = self._create_viz_types_buttons()
+
+        # Create vbox for shelves
+        shelves_view = self._create_shelves_controls()
+
+        # Create left pane with all controls
+        left_pane = w.VBox()
+        left_pane.children = [viz_types_hbox, shelves_view]
+
+        return left_pane
+
+    def _create_shelves_controls(self):
+        columns, types = self._find_columns_and_types(self.viz.data)
+        vbox = w.VBox()
+        if not hasattr(self.viz, 'encoding'):
+            setattr(self.viz, 'encoding', Encoding())
+        encoding = self._create_encoding_view(columns, types)
+        vbox.children = [encoding]
+        return vbox
+
+    def _create_viz_types_buttons(self):
+        hbox = w.HBox()
+
+        heading = w.HTML('Mark:', width='50px', height='32px')
+
+        # TODO: Could not express this method succinctly as on_press always ends up calling viz.text()
```

Comment:
This really bothered me, not being able to do without all the manual on_render methods below and whatnot, but as the comment states, I couldn't get on_press to have different viz_mark_method. It always ends up calling viz.text()

A better python programmer could probably tell me why that is.